### PR TITLE
ltx-2 audio decode: 2.24x on bh 2x4 — conv1d_depthwise op, conv blockings, channel-TP, polyphase, fused snake

### PR DESCRIPTION
**Pull request by smartonTT**
## Summary

Performance + correctness pass on the **LTX-2 audio decode** (mel-VAE → main vocoder → BWE generator). Adds the `conv1d_depthwise` op and routes the anti-alias filters through it, shape-tunes the fp32 conv3d blockings, channel-TPs the vocoders across the mesh, replaces the anti-alias `UpSample1d` with a polyphase form, and fuses `SnakeBeta`. fp32 throughout.

## Performance

Full audio decode (`decode_audio` = mel-VAE + main vocoder + BWE), **warm, real 6s clip (601 mel frames), bh 2x4 loudbox**:

| depthwise | warm wall |
|---|---|
| MAC (pre-conv1d baseline) | 2797 ms |
| **conv1d (this PR)** | **1250 ms — 2.24x** |

## Optimization defaults — what's active

| Optimization | Default | Notes |
|---|---|---|
| conv1d depthwise (`_USE_CONV1D_DEPTHWISE`) | **ON** | the 2.24x. MAC stays behind the flag as the bit-faithful baseline. |
| ups / AMP conv blockings, polyphase, SnakeBeta fusion | **ON** | always active. |
| main-vocoder channel-TP | **ON** in multi-axis configs (2x4 `sp1tp0`, Galaxy) | config-driven (both mesh axes > 1); single-axis meshes run T-parallel only. |
| BWE-generator channel-TP | **mesh-gated** | ON when channel factor > 2 (Galaxy 4x8/4x32); OFF on 2x4 (factor 2 — a measured +135 ms regression, gather tax > the split). Bit-identical to single-axis on real weights. `LTX_BWE_CHANNEL_TP=0/1` overrides. |

No optimization is unconditionally off. On the shipped 2x4 config everything above is active except BWE channel-TP, which is mesh-gated off there (a measured 2x4 regression) and turns on automatically for Galaxy — so nothing is silently disabled on the box it ships on.

## conv1d vs MAC

Measured on the real 6s clip (bh 2x4, warm); both vs a torch oracle (diffusers vocoder+BWE on CPU with the real checkpoint weights):

| Criterion | conv1d (default) | MAC (`_USE_CONV1D_DEPTHWISE=False`) |
|---|---|---|
| Full audio decode wall | **1250 ms (2.24x)** | 2797 ms |
| Fidelity vs torch oracle (typical 1s interval) | ~−18 dB | **~−23 dB** |
| Overall PCC vs torch | 0.9938 | **0.9961** |
| Bit-faithful to the sequential MAC reduction | no (reorders kaiser-sinc taps) | **yes** |
| Introduces a localized burst vs MAC | **no** (uniform ~−30 dB) | — |

**Decision:** ship conv1d (2.24x; introduces no localized static — the conv1d-vs-MAC difference is uniform). MAC is ~5 dB closer to ground truth, so flip `_USE_CONV1D_DEPTHWISE=False` if a listening A/B judges conv1d's coloration unacceptable. Note: both paths show a localized ~−12 dB divergence from torch at one ~1s interval of this clip — present in **both**, so it's a pipeline-numerics issue (not the depthwise choice), out of scope here.

## Changes
1. **`conv1d_depthwise` op** — standalone halo-aware depthwise 1D FIR op with shared scalar taps. By @kevinmi-TT (cherry-picked from `kevinmi/conv1d-depthwise`); see #46380.
2. **Route `depthwise_tap_filter` → `conv1d_depthwise`** — replaces the K-tap MAC fallback (and the `ttnn.conv1d` path that hung on T-sharded meshes).
3. **ups inner-conv blockings** in `_FP32_BLOCKINGS` — the 6 `ConvTranspose1d` inner convs fell to the `(32,32,1)` default (`T_out_block=1`); now shape-tuned.
4. **Retune AMP-block conv blockings** + fix dead `(64,48)` keys.
5. **Channel-TP on the second mesh axis** when both axes > 1. Adds `_pick_c_out_block_shard` so the conv3d device-op assertions hold for any shard size.
6. **Polyphase decomposition for the anti-alias `UpSample1d`** — 2 parallel length-(K/2) stride-1 convs vs zero-stuff + length-K depthwise + crop. Bit-equivalent; halves the anti-alias upsample.
7. **Fuse `SnakeBeta` via `ttnn.snake_beta`** — one tilize + fused op vs the 6-op eltwise chain.
8. **MAC depthwise fallback (`_USE_CONV1D_DEPTHWISE`) + mesh-gated BWE-generator channel-TP** — see the defaults table. `LTX_DUMP_AUDIO_LATENT` dumps the generated latent for offline audio-decode replay.
9. **Audio-decode regression + girl-clip tests** — `test_stage_c_audio_decode` (default-on: full VocoderWithBWE at 6s, per-0.5s-window sharded-vs-unsharded gate); `test_audio_decode_girl` (checkpoint-gated: the audio section alone with real weights, no transformer/gemma, per-1s-interval rmse/σ + dB + PCC vs a torch oracle, writes wavs).
10. **Per-conv / category / BWE profilers** (`prof_vocoder_forward.py`) for tuning, incl. `test_prof_bwe_per_conv` (per-conv `T_out_block` + tuned-or-not).

## Test plan
- [x] `test_audio_components_ltx` — stage A/B/C + `test_stage_c_audio_decode` pass; shard-vs-unsharded `max|Δ|` ~3.7e-4 < 5e-3 gate (2x4).
- [x] `test_audio_decode_girl` (with `LTX_CHECKPOINT`) — real 6s decode, warm 1250 ms (conv1d); per-1s error + PCC vs torch oracle.
- [x] `ttnn/.../test_conv1d_depthwise.py` — the op's unit test.

## Reproduce

The real girl-clip latent ships as a fixture (`tests/models/ltx/fixtures/girl_audio_latent.npy`, 39 KB fp16), so the audio decode runs on real content with only `LTX_CHECKPOINT` — no gemma, no gen:

```bash
export HF_HUB_OFFLINE=1 TT_DIT_CACHE_DIR=~/.cache/tt-dit
export LTX_CHECKPOINT=<path>/ltx-2.3-22b-distilled-1.1.safetensors

# per-1s error-to-signal (rmse/sigma + dB) and PCC vs the torch oracle for conv1d AND MAC,
# and write girl_audio_{conv1d,mac,torch}.wav:
AUDIO_OUT=/tmp/ab pytest models/tt_dit/tests/models/ltx/test_pipeline_ltx_distilled_av.py::test_audio_decode_girl -k bh_2x4sp1tp0 -s

# full-audio wall A/B (MAC vs conv1d):
for d in mac conv1d; do AUDIO_DEPTHWISE=$d \
  pytest models/tt_dit/tests/models/ltx/test_pipeline_ltx_distilled_av.py::test_audio_decode_girl -k bh_2x4sp1tp0 -s; done

# (optional) regenerate the fixture from a fresh gen (needs gemma):
GEMMA_PATH=google/gemma-3-12b-it-qat-q4_0-unquantized NO_PROMPT=1 LTX_DUMP_AUDIO_LATENT=/tmp/l.pt \
  pytest models/tt_dit/tests/models/ltx/test_pipeline_ltx_distilled_av.py::test_pipeline_av_fast -k bh_2x4sp1tp0 -s
# then: python -c "import torch,numpy as np; np.save('models/tt_dit/tests/models/ltx/fixtures/girl_audio_latent.npy', torch.load('/tmp/l.pt').numpy().astype(np.float16))"
```

Grep `AUDIO_GIRL` / `[oracle]` for the numbers.

## Follow-up
- **#46423** — Galaxy (4x8/4x32) scaling + tuning: T-pad inflation, validate the BWE channel-TP factor-4 crossover, conv3d blocking re-tune for Galaxy shard shapes.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:smarton/ltx-av-conv-perf)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:smarton/ltx-av-conv-perf)

### CODE REVIEW

File: models/tt_dit/tests/models/ltx/test_audio_components_ltx.py [hunk 1]
```diff
@@ -254,15 +254,25 @@ def _build_torch_stage_b(seed: int = 42):
     return model
 
 
-def _audio_parallel_config(mesh_shape: tuple[int, int]) -> ParallelFactor | None:
+def _audio_parallel_config(mesh_shape: tuple[int, int]):
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/tests/models/ltx/test_audio_components_ltx.py]: Fixed — restored `-> AudioTCParallelConfig | ParallelFactor | None` (folded into the channel-TP commit).

File: ttnn/cpp/ttnn/operations/experimental/conv1d_depthwise/device/kernels/compute_conv1d_depthwise.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    constexpr uint32_t block_num_tiles = block_w_tiles * block_h_tiles;
+
+    binary_op_init_common(tilized_cb, scalar_cb, out_cb);
+
+    for (uint32_t blk = 0; blk < num_blocks; ++blk) {
+        for (uint32_t tap = 0; tap < K; ++tap) {
+            compute_kernel_lib::tilize<block_w_tiles, act_cb, tilized_cb>(block_h_tiles);
+            mul_and_accumulate<block_num_tiles>(tilized_cb, scalar_cb, out_cb, tap);
+        }
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply ttnn/cpp/ttnn/operations/experimental/conv1d_depthwise/device/kernels/compute_conv1d_depthwise.cpp:76]: **Applied `Fp32Mode::Lossless`** to the per-tap tilize (commit 1). Measured on the vocoder: wall 356 ms and PCC 99.9868% / max|Δ| 2.14e-4 are bit-unchanged vs Fast, so it's free here — and it's the correct mode for the fp32 baseline (this op replaced an all-fp32 MAC path). conv1d_depthwise smoke suite green.

File: tests/ttnn/unit_tests/operations/conv/test_conv1d_depthwise.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+# Smoke shapes: small, fast, single-device. Covers C padding (16/24/48), strides, K.
+SMOKE = [
+    (1, 256, 64, 12, 1),
+    (1, 256, 64, 12, 2),
+    (1, 512, 32, 12, 1),
+    (1, 512, 16, 12, 2),
+    (1, 300, 24, 12, 1),
+    (1, 320, 48, 4, 1),
+    (1, 600, 128, 29, 1),
+]
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply tests/ttnn/unit_tests/operations/conv/test_conv1d_depthwise.py:56]: Update: addressed in this PR — added C=1 and C=2 smoke cases (small T) exercising the C<TILE padding path. Both pass (C=2 PCC 0.9999997).

File: tests/ttnn/unit_tests/operations/conv/test_conv1d_depthwise.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    (32, 28811, 12, 2),
+    (32, 57605, 4, 1),
+    (16, 115241, 12, 1),
+    (16, 57611, 12, 2),
+]
+HANNUP = [(1, 19228, 29, 1), (2, 19228, 29, 1)]
+
+# (C, T_pad, K, stride) -> single-device, B=1. Tail (T_pad > 30k) marked slow.
+PROD = [(c, t, k, s) for (c, t, k, s) in (STAGE_B + STAGE_C)] + [
+    pytest.param(c, t, k, s, marks=pytest.mark.skip(reason="C<TILE handled by smoke; HannUp C=1/2 padding case"))
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply tests/ttnn/unit_tests/operations/conv/test_conv1d_depthwise.py]: Update: fixed in this PR — added the C=1/2 smoke cases (so the claim is now true) and rewrote the skip reason to 'C=1/2 padding path covered by smoke; HannUp T=19228 tail redundant with prod-tail'.

**Comment by mstaletovicTT** [comment]
Who are the codeowners for these new operations ? 

Also, this code looks generally extremely familiar with the code for the already existing conv1d depthwise path. I see that you had OOM issues, but do you think that at least some part of the code could be reused ? Especially if you intend to put the Convolution Team as CODEOWNERS. Are there any plans to merge this with the existing paths ?

**Comment by mstaletovicTT** [comment]
Also, would it be possible to get some kind of unit test repro for when you  had issues with using conv1d (topology, metal checkout, unit test) .

**Comment by smartonTT** [comment]
@mstaletovicTT — thanks, and a few of these are really for @kevinmi-TT since the op itself isn't mine.

**Authorship / CODEOWNERS / merging with the existing depthwise path:** commit 1 (`conv1d_depthwise`) is cherry-picked from @kevinmi-TT's `kevinmi/conv1d-depthwise` (#46380) — I only adopted it in the vocoder. You're right that it overlaps the existing depthwise conv1d code; whether to unify the paths, and ownership (Conv team makes sense — there's no `CODEOWNERS` entry for `ttnn/.../experimental/conv1d_depthwise/` today), is @kevinmi-TT's + the Convolution team's call, not something I should decide unilaterally in this PR. @kevinmi-TT — can you weigh in on dedup/merge plans and CODEOWNERS for these files?

**Why the separate op (the conv1d repro you asked for):** in the vocoder the anti-alias resample filters are *depthwise* with shared scalar taps — C=1/2 (padded to 32), kernel K≤12, and T grows large after the upsample stages — run on a **T-sharded** mesh. The stock `ttnn.conv1d` path OOM'd L1 / hung (surfaced as a `synchronize_device` timeout) at those shapes, which is what pushed us to a dedicated op. I'll put together a minimal standalone repro (topology + exact shapes + the failing call) and post it here so it's reproducible; @kevinmi-TT may already have one from #46380.

Separately: the pre-commit + check-spdx-licenses failures are fixed in the latest push (missing SPDX header on the prof file, a dead `_STATIC_SR`, black/autoflake).

**Comment by smartonTT** [comment]
@mstaletovicTT — the conv1d repro is up on the `ltx-perf` branch, right alongside the op (which is on ltx-perf too) and the rest of the audio work: `tests/ttnn/unit_tests/operations/conv/test_conv1d_depthwise.py::test_stock_conv1d_l1_oom_vs_depthwise`.

- **Topology:** single Blackhole device, `l1_small_size=32768` (the vocoder's setting).
- **Shape:** a real STAGE_C anti-alias upsample tail — depthwise, C=64, T_pad=28841, K=12, stride=1.
- **Result:** stock `ttnn.conv1d(groups=C, shard_layout=HEIGHT_SHARDED)` throws `TT_THROW @ program.cpp:1470` (static circular buffers exceed the 1.5 MB L1 — it buffers the whole sequence), while `ttnn.experimental.conv1d_depthwise` runs the same shape and matches the depthwise golden. On a T-sharded mesh the same path hung instead of throwing.

This PR (#46395) carries the op to `main`; the repro currently lives on `ltx-perf` (where the audio work is being integrated). Say the word if you'd like it cherry-picked into this PR as well.

**Comment by smartonTT** [comment]
@mstaletovicTT — having read both kernels: you're right, it's very close to the existing path. Concretely, the compute kernel (`compute_conv1d_depthwise.cpp`) mirrors `mul_and_accumulate_block` in `conv2d/device/kernels/compute_depthwise_conv1d.cpp` — tilize block → per-tap `mul_tiles` → `tap>0` dest-reuse ELWADD → single pack. That part is duplicative and could be unified.

What's genuinely new here, and why it's separate: (1) the weight is K **shared scalars** (one resident tile, broadcast over channels), since the vocoder anti-alias filter is identical per channel — not per-channel weight tiles; (2) a **row-major, T-tile-row work-split** that streams the sequence across cores instead of the conv2d depthwise path's HEIGHT_SHARDED reader, which buffers the full sequence in L1 and OOMs at the long T after the upsample stages (repro: `test_stock_conv1d_l1_oom_vs_depthwise`); (3) fp32 + Lossless tilize.

So the realistic dedup is: share the compute kernel, and add the streaming-schedule + scalar-weight as a mode on the existing depthwise conv1d rather than a standalone op. That's a Conv-team call (@kevinmi-TT authored the op, #46380) — flagging it as a real follow-up, not something to resolve here.

**Comment by smartonTT** [comment]
Closing — superseded by the `ltx-perf` integration branch, which now carries all of this audio work (conv1d_depthwise adoption, conv blockings, channel-TP, polyphase, fused SnakeBeta) plus the later additions (MAC fallback flag, mesh-gated BWE channel-TP, torch-oracle regression tests, the real girl-clip fixture, and the conv1d L1-OOM repro). It will reach `main` via that branch rather than this PR. The conv1d_depthwise op itself continues via #46380.
